clients/pkg/promtail/targets/docker/target.go

```diff
@@ -0,0 +1,212 @@
+		metrics:   metrics,
+
+		ctx:     ctx,
+		cancel:  cancel,
+		client:  client,
+		running: atomic.NewBool(false),
+	}
+	t.start()
+	return t, nil
+}
+
+func (t *Target) start() {
+	t.running.Store(true)
+
+	opts := docker_types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     true,
+		Timestamps: true,
+		// Since: "todo", detemer since from position
+	}
+
+	logs, err := t.client.ContainerLogs(t.ctx, t.config.ContainerName, opts)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "could not fetch logs for container", "container", t.config.ContainerName, "err", err)
+		t.err = err
+		return
+	}
+
+	out := make(chan frame)
+	dstout := stdoutWriter{out: out}
+	dsterr := stdoutWriter{out: out}
+
+	// Start transfering
+	go func() {
+		t.wg.Add(1)
+		defer func() {
+			t.wg.Done()
+			close(out)
+		}()
+
+		written, err := stdcopy.StdCopy(dstout, dsterr, logs)
+		if err != nil {
+			level.Error(t.logger).Log("msg", "could not transfer logs", "written", written, "err", err)
+		} else {
+			level.Info(t.logger).Log("msg", "finished transfering logs", "written", written)
+		}
+	}()
+
+	// Start processing
+	go func() {
+		t.wg.Add(1)
+		defer func() {
+			t.wg.Done()
+			t.running.Store(false)
+			logs.Close()
+		}()
+
+		for t.ctx.Err() == nil {
+			select {
+			case f := <-out:
+				if len(f.line) == 0 {
+					continue
+				}
+
+				ts, line, err := extractTs(f.line)
+				if err != nil {
+					level.Error(t.logger).Log("msg", "could not extract timestamp", "err", err)
+					t.metrics.dockerErrors.Inc()
+				}
+				t.handler.Chan() <- api.Entry{
+					Labels: t.config.Labels.Clone(),
+					Entry: logproto.Entry{
+						Timestamp: ts,
+						Line:      line,
+					},
+				}
+				t.metrics.dockerEntries.Inc()
+			case <-t.ctx.Done():
+				{
+				}
+			}
+		}
+	}()
+}
+
+func extractTs(line string) (time.Time, string, error) {
+	pair := strings.SplitN(line, " ", 2)
+	if len(pair) != 2 {
+		return time.Now(), line, fmt.Errorf("Could not find timestamp in '%s'", line)
+	}
+	ts, err := time.Parse("2006-01-02T15:04:05.999999999Z07:00", pair[0])
+	if err != nil {
+		return time.Now(), line, fmt.Errorf("Could not parse timestamp from '%s': %w", pair[0], err)
+	}
+	return ts, pair[1], nil
+}
+
+func (t *Target) Stop() {
+	t.cancel()
+	t.wg.Wait()
+	t.handler.Stop()
+}
+
+func (t *Target) Type() target.TargetType {
+	return target.DockerTargetType
+}
+
+func (t *Target) Ready() bool {
+	return t.running.Load()
+}
+
+func (t *Target) DiscoveredLabels() model.LabelSet {
+	return nil // TODO
+}
+
+func (t *Target) Labels() model.LabelSet {
+	return t.config.Labels
+}
+
+// Details returns target-specific details.
+func (t *Target) Details() interface{} {
+	return map[string]string{}
+}
+
+type frame struct {
+	stream stdcopy.StdType
+	line   string
+}
+
+type stdoutWriter struct {
+	out chan frame
+}
+
+func (w stdoutWriter) Write(p []byte) (n int, err error) {
+	f := frame{stream: stdcopy.Stdout, line: string(p)}
+	w.out <- f
```

Comment:
This call may block forever, if goroutine reading the channel doesn't read it anymore (if `t.ctx` is done, goroutine will stop reading the channel).

Better alternative may be to use `io.Pipe()`, which returns both reader and writer. Writer should be supplied to `stdcopy.StdCopy` call, while goroutine reading the reader will do line-splitting and use the log. If any of the pipe's ends is closed prematurely, the other goroutine will get error, instead of being blocked.

Reply:
(Two `os.Pipe()` calls are needed, one for stdout, one for stderr)

Reply:
Thanks for taking a look.

> while goroutine reading the reader will do line-splitting and use the log.
Are the Reader and Writer "threadsafe"? Is it safe to write in one goroutine and read in another?

Reply:
Ah the docs answer that question :slightly_smiling_face: 

> It is safe to call Read and Write in parallel with each other or with Close.

Perfect. That would be much simpler.

Reply:
Hm, how would I react to a `t.ctx.Done` though?